content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,242 @@
+    image: valkey/valkey:latest
+    hostname: valkey1
+    command: valkey-server --port 6379 --requirepass ${VALKEY_PASSWORD} --io-threads ${IO_THREADS}
+    volumes:
+      - ./data:/data
+    network_mode: host
+
+volumes:
+  data:
+    driver: local
+```
+
+Since we are exposing this to our internal network we will create a password for our default user. I used `head -16 /dev/urandom | openssl sha1` to generate a random password. Because of how fast Valkey can process requests a brute force attack could try hundereds of thousands of passwords per second. After generating that password I put it in a `.env` file in the same directory as our docker compose. 
+
+```bash
+#.env
+VALKEY_PASSWORD=e41fb9818502071d592b36b99f63003019861dad
+NODE_IP=<VALKEY SERVER IP>
+IO_THREADS=1
+```
+
+Now by running `docker compose -f valkey.yaml up -d` Valkey server will start with password we set! 
+
+## Baseline test
+
+Now we are ready to do some baseline testing. We will log into the benchmarking server. If you haven't installed redis-benchmark yet you can do so with `sudo apt install redis-tools`.  
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a <PASSWORD FROM .env> --threads 5 -h 10.0.1.136
+```
+
+Test breakdown:
+- `-n` - this will run 1,000,000 operations using the commands in -t
+- `-t` - will run the set and get tests
+- `-P` - this specifies that we would like the tests to use 16 pipelines (send 16 operations per request). 
+- `-q` - silences the output to show only the final results
+- `-a` - use the specified password
+- `-h` - run the test against the specified host
+- `--threads` - how many threads to generate test data from
+
+Honestly, I was astonished by the first set of results I got. Sure, I expected Valkey to be fast but this speed from a single board computer?? ON A SINGLE THREAD?? It's amazing. 
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad --threads 5 -h 10.0.1.136
+SET: 173040.33 requests per second, p50=4.503 msec                    
+GET: 307031.00 requests per second, p50=2.455 msec
+```
+
+Between the two tests we averaged 240,000 requests per second.  
+
+## Raising the CPU clock speed
+
+Since Valkey is a single threaded application it makes sense that higher clock speeds would lead to more performance. I don't expect most people will overclock their servers in production. Different servers may be available with different CPU clock speeds. 
+
+**Note:** Clock speeds generally are only comparable between CPU's with a similar architecture. For example, you could reasonably compare clock speeds between an 12th generation Intel i5 and a 12th generation Intel i7. If the 12th gen i7 had a max clock speed of 5Ghz that doesn't necessarily mean it will be slower than a AMD Ryzen 9 9900X clocked at 5.6Ghz. 
+
+If you're following along on a Pi of your own I've outlined the steps to overclock your CM4 below. Otherwise you can skip to the results section below.
```

Comment:
```suggestion
If you're following along on a Pi of your own I've outlined the steps to overclock your CM4 below. Otherwise, you can skip to the results section below.
```